### PR TITLE
buildMavenPackage: Add `overrideMavenAttrs` function and `buildOffline` documentation

### PR DESCRIPTION
**Pull request by tricktron**
## Description of changes

This pr does the following:
-  Adds a `overrideMavenAttrs` function to make it possible to override attributes of the `buildMavenPackage` function.
- Documents the `overrideMavenAttrs` function.
- Documents the use case of the `buildOffline = true` argument for the `buildMavenPackage` function.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by tricktron** [comment]
@ofborg build lemminx dbeaver java-language-server h2 scenebuilder schemaspy

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (fricklerhandwerk) [reply pkgs/development/tools/build-managers/apache-maven/default.nix:40]: @roberth argues we may want a more specific names for the argument, such as `mavenArgs`.

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]
Discussed at #ZurichZHF

File: doc/languages-frameworks/maven.section.md
Comment (fricklerhandwerk) [inline doc/languages-frameworks/maven.section.md:80]:
````suggestion
```
overrideMavenAttrs :: (AttrSet -> Derivation) | ((AttrSet -> Attrset) -> Derivation) -> Derivation
```

The output of `buildMavenPackage` has an `overrideMavenAttrs` attribute, which is a function that takes either
- any subset of the attributes that can be passed to `buildMavenPackage`

  or
- a function that takes the argument passed to the previous invocation of `buildMavenPackage` (conventionally called `old`) and returns an attribute set that can be passed to `buildMavenPackage`

and returns a derivation that builds a Maven package based on the old and new arguments merged.

This is similar to [](#sec-pkg-overrideAttrs), but notably does not allow accessing the final value of the argument to `buildMavenPackage`.

:::{.example}

Use `overrideMavenAttrs` to build `jd-cli` version 1.2.0 and disable some flaky test:

```nix
jd-cli.overrideMavenAttrs (old: rec {
  version = "1.2.0";
  src = fetchFromGitHub {
    owner = old.src.owner;
    repo = old.src.repo;
    rev = "${old.pname}-${version}";
    # old source hash of 1.2.0 version
    hash = "sha256-US7j6tQ6mh1libeHnQdFxPGoxHzbZHqehWSgCYynKx8=";
  };

  # tests can be disabled by prefixing it with `!`
  # see Maven documentation for more details:
  # https://maven.apache.org/surefire/maven-surefire-plugin/examples/single-test.html#Multiple_Formats_in_One
  mvnParameters = lib.escapeShellArgs [
    "-Dsurefire.failIfNoSpecifiedTests=false"
    "-Dtest=!JavaDecompilerTest#basicTest,!JavaDecompilerTest#patternMatchingTest"
  ];

  # old mvnHash of 1.2.0 maven dependencies
  mvnHash = "sha256-N9XC1pg6Y4sUiBWIQUf16QSXCuiAPpXEHGlgApviF4I=";
});
```
:::
````

File: doc/languages-frameworks/maven.section.md
Comment (fricklerhandwerk) [inline doc/languages-frameworks/maven.section.md:84]:
```suggestion
By default, `buildMavenPackage` does the following:
```

File: doc/languages-frameworks/maven.section.md
Comment (fricklerhandwerk) [inline doc/languages-frameworks/maven.section.md:89]:
```suggestion
1. Run `mvn package -Dmaven.repo.local=$out/.m2 ${mvnParameters}` in the
   `fetchedMavenDeps` [fixed-output derivation](https://nixos.org/manual/nix/stable/glossary.html#gloss-fixed-output-derivation).
2. Run `mvn package -o -nsu "-Dmaven.repo.local=$mvnDeps/.m2"
   ${mvnParameters}` again in the main derivation.
```

File: doc/languages-frameworks/maven.section.md
Comment (fricklerhandwerk) [inline doc/languages-frameworks/maven.section.md:94]:
```suggestion
As a result, tests are run twice.
This also means that a failing test will trigger a new attempt to realise the fixed-output derivation, which in turn downloads all dependencies again.
For bigger Maven projects, this might lead to a long feedback cycle. 
```

File: doc/languages-frameworks/maven.section.md
Comment (fricklerhandwerk) [inline doc/languages-frameworks/maven.section.md:101]:
```suggestion
Use `buildOffline = true` to change the behaviour of `buildMavenPackage to the following:
1. Run `mvn de.qaware.maven:go-offline-maven-plugin:1.2.8:resolve-dependencies
   -Dmaven.repo.local=$out/.m2 ${mvnDepsParameters}` in the fixed-output derivation.
2. Run `mvn package -o -nsu "-Dmaven.repo.local=$mvnDeps/.m2"
   ${mvnParameters}` in the main derivation.
```

File: doc/languages-frameworks/maven.section.md
Comment (fricklerhandwerk) [inline doc/languages-frameworks/maven.section.md:105]:
```suggestion
As a result, all dependencies are downloaded in step 1 and the tests are executed in step 2.
A failing test only triggers a rebuild of step 2 as it can reuse the dependencies of step 1 because they have not changed.
```

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]

**Comment by tricktron** [comment]
@roberth Any idea on how to make this work with finalAttrs?
E.g.
```nix
jd-cli.overrideMavenAttrs (finalAttrs: old: rec {
...
```

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (roberth) [inline pkgs/development/tools/build-managers/apache-maven/default.nix:54]:
```suggestion
        drv // {
            overrideMavenAttrs = newArgs: makeOverridableMavenPackage mavenRecipe (overrideWith newArgs);
        };
```

IIUC, the recipe function always returns attrs.
Code like this is confusing enough even without dead code.

**Comment by roberth** [comment]
> @roberth Any idea on how to make this work with finalAttrs? E.g.
> 
> ```nix
> jd-cli.overrideMavenAttrs (finalAttrs: old: rec {
> ...
> ```

I'm not 100% sure what it should mean, and while it could be done, the "architecture" - functions composed in sequence, but made overridable - does not make this easy, and requires that the logic to make that work have knowledge about all other relevant overriding functions, such as `overrideAttrs` or even others introduced via `passthru`.
For this reason I'm proposing to change the architecture in
- https://github.com/NixOS/nixpkgs/issues/273815

I'm sure you could make it work for limited use cases, e.g. initially not supporting `overrideMavenAttrs` after `overrideAttrs` has been called; I just think effort is better spent on solving the root of the problem.

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (fricklerhandwerk) [reply pkgs/development/tools/build-managers/apache-maven/default.nix:54]: Yes, we actually already discussed this during review when we met.

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (tricktron) [reply pkgs/development/tools/build-managers/apache-maven/default.nix]: Done.

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (tricktron) [reply pkgs/development/tools/build-managers/apache-maven/default.nix]: Done.

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (tricktron) [reply pkgs/development/tools/build-managers/apache-maven/default.nix]: Done

File: pkgs/development/tools/build-managers/apache-maven/default.nix
Comment (tricktron) [reply pkgs/development/tools/build-managers/apache-maven/default.nix]: Done.

**Comment by tricktron** [comment]
@roberth @fricklerhandwerk I cleaned up the pr for you guys. Unfortunately, I cannot attend the second ZHF day, so I won't be onsite for more interesting discussions.

**Comment by tomodachi94** [comment]
Also, @tricktron and company: It would be awesome to see you in [#jvm:nixos.org](https://matrix.to/#/#jvm:nixos.org) if you're interested in helping maintain Nixpkgs Java ecosystem at all 🙂

**Comment by tricktron** [comment]
> This is interesting. I'm generally not opposed to this, but I think the plan should be to eventually remove `buildMavenPackage` in favor of setup hooks. @NixOS/java: thoughts?

I have discussed the same thing with @wegank today and I agree that setup hooks might be an even better improvement.

So we could first merge that in and setup the switch to setup hooks in a future pr. I am fine with contributing to that too.

What do you think?

**Comment by tomodachi94** [comment]
> What do you think?

I think that would be ideal in this situation. I'll take a look at this in a moment (if I can wrap my head around how overrides work behind the scenes :wink:).

**Comment by tricktron** [comment]
@fricklerhandwerk Could you unblock merging this and/or update your needed points?

**Review by fricklerhandwerk — APPROVED** [review]

**Review by tomodachi94 — APPROVED** [review]
Lgtm, thank you 💖 If you're interested in maintaining this package, I encourage you to add yourself to `meta.maintainers`.

**Comment by tricktron** [comment]
> Lgtm, thank you 💖 If you're interested in maintaining this package, I encourage you to add yourself to `meta.maintainers`.

Sure, I added myself to the maintainer list. I may also be interested to join the Nixos Java team in the future. For now, I'll try to check in from time to time to https://matrix.to/#/#jvm:nixos.org but probably more as a silent reader if that is ok:smiley:.

**Review by tomodachi94 — APPROVED** [review]
Sounds good, thank you ❤️ If you have something to say, feel free to provide input when there's discussion in the channel.

**Comment by TomaSajt** [comment]
Just saw that this happened, so I quickly threw together an alternative: https://github.com/NixOS/nixpkgs/pull/359573

**Pull request merged**
